File: packages/swift-agent-protocol/Sources/OpenPawProtocol/Inbox.swift
import Foundation

// MARK: - Inbox vocabulary

public enum InboxCategory: String, Codable, Sendable, Hashable, CaseIterable {
    case permission
    case question
    case plan
    case toolFailure = "tool_failure"
    case completion
    case contextWarning = "context_warning"
    case rateLimit = "rate_limit"
    case backgroundJob = "background_job"
}

public enum InboxStatus: String, Codable, Sendable, Hashable, CaseIterable {
    case pending
    case resolved
    case dismissed
    case expired
}

/// Whether a client may render a single tap approval, or must first expand the full
/// detail. `.requiresDetailExpansion` carries the specific triggers so the UI can name
/// them instead of showing a generic warning.
public enum ApprovalGate: Sendable, Hashable {
    case oneTap
    case requiresDetailExpansion(reasons: [String])

    public var isOneTap: Bool {
        if case .oneTap = self { return true }
        return false
    }

    public var reasons: [String] {
        if case .requiresDetailExpansion(let reasons) = self { return reasons }
        return []
    }
}

// MARK: - Inbox item

public struct InboxItem: Codable, Sendable, Hashable, Identifiable {
    public let id: InboxID
    public let sessionID: SessionID
    public let agent: AgentKind
    public let category: InboxCategory
    public let title: String
    public let detail: String?
    public let command: String?
    public let risk: Risk?
    public let actions: [ActionID]
    /// Underlying `permission.requested` / `question.requested` request id, used to
    /// correlate a later `permission.resolved` back to this item.
    public let requestID: String?
    /// One-time bearer required by `POST /v1/inbox/{id}/resolve`. Never present in a
    /// push payload; only fetched over the authenticated tunnel.
    public var actionToken: String?
    public let createdAt: Date
    public let expiresAt: Date?
    public var status: InboxStatus
    public var resolution: String?
    public let sourceEventID: EventID

    public init(
        id: InboxID,
        sessionID: SessionID,
        agent: AgentKind,
        category: InboxCategory,
        title: String,
        detail: String? = nil,
        command: String? = nil,
        risk: Risk? = nil,
        actions: [ActionID],
        requestID: String? = nil,
        actionToken: String? = nil,
        createdAt: Date,
        expiresAt: Date? = nil,
        status: InboxStatus,
        resolution: String? = nil,
        sourceEventID: EventID
    ) {
        self.id = id
        self.sessionID = sessionID
        self.agent = agent
        self.category = category
        self.title = title
        self.detail = detail
        self.command = command
        self.risk = risk
        self.actions = actions
        self.requestID = requestID
        self.actionToken = actionToken
        self.createdAt = createdAt
        self.expiresAt = expiresAt
        self.status = status
        self.resolution = resolution
        self.sourceEventID = sourceEventID
    }

    enum CodingKeys: String, CodingKey {
        case id
        case sessionID = "session_id"
        case agent
        case category
        case title
        case detail
        case command
        case risk
        case actions
        case requestID = "request_id"
        case actionToken = "action_token"
        case createdAt = "created_at"
        case expiresAt = "expires_at"
        case status
        case resolution
        case sourceEventID = "source_event_id"
    }

    public init(from decoder: any Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        id = try container.decode(InboxID.self, forKey: .id)
        sessionID = try container.decode(SessionID.self, forKey: .sessionID)
        agent = try container.decode(AgentKind.self, forKey: .agent)
        category = try container.decode(InboxCategory.self, forKey: .category)
        title = try container.decode(String.self, forKey: .title)
        detail = try container.decodeIfPresent(String.self, forKey: .detail)
        command = try container.decodeIfPresent(String.self, forKey: .command)
        risk = try container.decodeIfPresent(Risk.self, forKey: .risk)
        actions = try container.decodeIfPresent([ActionID].self, forKey: .actions) ?? []
        requestID = try container.decodeIfPresent(String.self, forKey: .requestID)
        actionToken = try container.decodeIfPresent(String.self, forKey: .actionToken)
        createdAt = try container.decode(Date.self, forKey: .createdAt)
        expiresAt = try container.decodeIfPresent(Date.self, forKey: .expiresAt)
        status = try container.decode(InboxStatus.self, forKey: .status)
        resolution = try container.decodeIfPresent(String.self, forKey: .resolution)
        sourceEventID = try container.decode(EventID.self, forKey: .sourceEventID)
    }

    /// Gate a client MUST honour before offering an approval action.
    public var approvalGate: ApprovalGate {
        guard let risk, risk.requiresDetailExpansion else { return .oneTap }
        return .requiresDetailExpansion(reasons: risk.reasons)
    }

    /// True once `expiresAt` has passed relative to `now`.
    public func hasExpired(at now: Date) -> Bool {
        guard let expiresAt else { return false }
        return now >= expiresAt
    }
}

// MARK: - Projection

/// Projects normalized events into actionable inbox items. Mirrors the host's
/// `InboxItem::from_event` so an offline client derives the same list from a replayed
/// event stream that the host would have served.
public struct InboxProjection: Sendable {
    /// Threshold above which a context update becomes a warning the user must see.
    public static let contextWarningPercent = 85.0
    /// Threshold above which a usage update becomes a rate limit warning.
    public static let rateLimitWarningPercent = 90.0

    public init() {}

    /// Returns `nil` for events that carry no decision or notification.
    public static func from(event: Event) -> InboxItem? {
        let id = InboxID(event: event.eventID)

        func item(
            category: InboxCategory,
            title: String,
            detail: String? = nil,
            command: String? = nil,
            risk: Risk? = nil,
            actions: [ActionID],
            requestID: String? = nil,
            expiresAt: Date? = nil
        ) -> InboxItem {
            InboxItem(
                id: id,
                sessionID: event.sessionID,
                agent: event.agent,
                category: category,
                title: title,
                detail: detail,
                command: command,
                risk: risk,
                actions: actions,
                requestID: requestID,
                actionToken: nil,
                createdAt: event.timestamp,
                expiresAt: expiresAt,
                status: .pending,
                resolution: nil,
                sourceEventID: event.eventID
            )
        }

        switch event.body {
        case .permissionRequested(let payload):
            return item(
                category: .permission,
                title: payload.summary,
                detail: payload.command ?? payload.summary,
                command: payload.command,
                risk: payload.risk,
                actions: payload.actions,
                requestID: payload.requestID,
                expiresAt: payload.expiresAt
            )

        case .questionRequested(let payload):
            return item(
                category: .question,
                title: payload.question,
                detail: payload.choices.isEmpty ? nil : payload.choices.joined(separator: ", "),
                actions: [.answer],
                requestID: payload.requestID
            )

        case .planCreated(let payload), .planUpdated(let payload):
            let base = payload.title ?? "Plan"
            let title = "\(base) (\(payload.completedCount)/\(payload.steps.count) steps complete)"
            let detail =
                payload.steps.isEmpty
                ? nil
                : payload.steps.map { "\($0.status.marker) \($0.title)" }.joined(separator: "\n")
            return item(category: .plan, title: title, detail: detail, actions: [.acknowledge])

        case .toolFailed(let payload):
            return item(
                category: .toolFailure,
                title: "Tool call \(payload.callID) failed",
                detail: payload.error,
                actions: [.acknowledge]
            )

        case .agentCompleted(let payload):
            return item(
                category: .completion,
                title: payload.title ?? "Agent completed",
                detail: payload.reason,
                actions: [.acknowledge]
            )

        case .agentFailed(let payload):
            return item(
                category: .toolFailure,
                title: payload.title ?? "Agent failed",
                detail: payload.reason,
                actions: [.acknowledge]
            )

        case .contextUpdated(let payload):
            guard payload.percentUsed >= contextWarningPercent else { return nil }
            return item(
                category: .contextWarning,
                title: "Context window \(rounded(payload.percentUsed))% used",
                detail: "\(payload.usedTokens) of \(payload.maxTokens) tokens used",
                actions: [.acknowledge]
            )

        case .usageUpdated(let payload):
            guard let percent = payload.rateLimitPercent, percent >= rateLimitWarningPercent else {
                return nil
            }
            return item(
                category: .rateLimit,
                title: "Rate limit \(rounded(percent))% consumed",
                detail: payload.rateLimitResetsAt.map { "resets at \(RFC3339.string(from: $0))" },
                actions: [.acknowledge]
            )

        case .agentStarted, .agentWorking, .turnStarted, .turnDelta, .turnCompleted,
            .toolStarted, .toolOutput, .toolCompleted, .permissionResolved, .questionAnswered,
            .fileRead, .fileCreated, .fileModified, .fileDeleted, .unsupported:
            return nil
        }
    }

    /// Half away from zero rounding, matching Rust's `f64::round`.
    private static func rounded(_ value: Double) -> Int64 {
        Int64(value.rounded(.toNearestOrAwayFromZero))
    }
}

File: host/crates/openpaw-protocol/src/inbox.rs
use serde::{Deserialize, Serialize};
use time::OffsetDateTime;

use crate::agent::AgentKind;
use crate::event::{Body, Event};
use crate::ids::{EventId, InboxId, SessionId};
use crate::payload::{ActionId, Plan, PlanStepStatus};
use crate::risk::Risk;
use crate::wire_enum::wire_enum;

wire_enum! {
    /// Why an inbox item exists.
    pub enum InboxCategory {
        /// The agent is blocked waiting for approval.
        Permission = "permission",
        /// The agent is blocked waiting for an answer.
        Question = "question",
        /// A plan was published or changed.
        Plan = "plan",
        /// A tool call or the agent itself failed.
        ToolFailure = "tool_failure",
        /// The agent finished its turn.
        Completion = "completion",
        /// The context window is nearly full.
        ContextWarning = "context_warning",
        /// The provider rate limit window is nearly exhausted.
        RateLimit = "rate_limit",
        /// A long running background job changed state.
        BackgroundJob = "background_job",
    }
}

wire_enum! {
    /// Lifecycle of an inbox item.
    pub enum InboxStatus {
        /// Awaiting a decision.
        Pending = "pending",
        /// A decision was recorded.
        Resolved = "resolved",
        /// Dismissed without a decision.
        Dismissed = "dismissed",
        /// The window for deciding closed.
        Expired = "expired",
    }
}

/// Context occupancy at or above this percentage raises a warning item.
const CONTEXT_WARNING_PERCENT: f64 = 85.0;

/// Rate limit consumption at or above this percentage raises a warning item.
const RATE_LIMIT_WARNING_PERCENT: f64 = 90.0;

/// An actionable projection of one event.
///
/// Actions carry a one-time token minted by the host; a push notification alone
/// is never sufficient to authorize a decision, which is why `action_token` is
/// always `None` here.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct InboxItem {
    /// Content addressed identifier, derived from `source_event_id`.
    pub id: InboxId,
    /// Session the item belongs to.
    pub session_id: SessionId,
    /// Agent that raised it.
    pub agent: AgentKind,
    /// Why the item exists.
    pub category: InboxCategory,
    /// One line notification-ready summary.
    pub title: String,
    /// Full detail the client must reveal before allowing a risky approval.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub detail: Option<String>,
    /// Raw command line when the underlying event carried one.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub command: Option<String>,
    /// Classification of the underlying operation.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub risk: Option<Risk>,
    /// Decisions a device may take.
    #[serde(default)]
    pub actions: Vec<ActionId>,
    /// One-time bearer for `POST /v1/inbox/{id}/resolve`, minted by the host.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub action_token: Option<String>,
    /// Underlying permission or question request id, so a later
    /// `permission.resolved` can be correlated back to this item.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub request_id: Option<String>,
    /// Timestamp of the source event.
    #[serde(with = "time::serde::rfc3339")]
    pub created_at: OffsetDateTime,
    /// Point after which the item can no longer be resolved.
    #[serde(
        default,
        with = "time::serde::rfc3339::option",
        skip_serializing_if = "Option::is_none"
    )]
    pub expires_at: Option<OffsetDateTime>,
    /// Lifecycle state.
    pub status: InboxStatus,
    /// Human readable outcome once resolved.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub resolution: Option<String>,
    /// Event this item was projected from.
    pub source_event_id: EventId,
}

/// The category-specific part of a projection.
struct Projection {
    category: InboxCategory,
    title: String,
    detail: Option<String>,
    command: Option<String>,
    risk: Option<Risk>,
    actions: Vec<ActionId>,
    request_id: Option<String>,
    expires_at: Option<OffsetDateTime>,
}

impl Projection {
    /// A purely informational projection: acknowledge and move on.
    fn informational(category: InboxCategory, title: String, detail: Option<String>) -> Projection {
        Projection {
            category,
            title,
            detail,
            command: None,
            risk: None,
            actions: vec![ActionId::Acknowledge],
            request_id: None,
            expires_at: None,
        }
    }
}

impl InboxItem {
    /// Projects an event into an inbox item, or `None` when the event needs no
    /// human decision and no notification.
    pub fn from_event(event: &Event) -> Option<InboxItem> {
        let projection = project(event)?;
        Some(InboxItem {
            id: InboxId::derive(&event.event_id),
            session_id: event.session_id.clone(),
            agent: event.agent,
            category: projection.category,
            title: projection.title,
            detail: projection.detail,
            command: projection.command,
            risk: projection.risk,
            actions: projection.actions,
            action_token: None,
            request_id: projection.request_id,
            created_at: event.timestamp,
            expires_at: projection.expires_at,
            status: InboxStatus::Pending,
            resolution: None,
            source_event_id: event.event_id.clone(),
        })
    }
}

fn project(event: &Event) -> Option<Projection> {
    match &event.body {
        Body::PermissionRequested(payload) => Some(Projection {
            category: InboxCategory::Permission,
            title: payload.summary.clone(),
            detail: Some(
                payload
                    .command
                    .clone()
                    .unwrap_or_else(|| payload.summary.clone()),
            ),
            command: payload.command.clone(),
            risk: Some(payload.risk.clone()),
            actions: payload.actions.clone(),
            request_id: Some(payload.request_id.clone()),
            expires_at: payload.expires_at,
        }),
        Body::QuestionRequested(payload) => Some(Projection {
            category: InboxCategory::Question,
            title: payload.question.clone(),
            detail: if payload.choices.is_empty() {
                None
            } else {
                Some(payload.choices.join(", "))
            },
            command: None,
            risk: None,
            actions: vec![ActionId::Answer],
            request_id: Some(payload.request_id.clone()),
            expires_at: None,
        }),
        Body::PlanCreated(plan) | Body::PlanUpdated(plan) => Some(Projection::informational(
            InboxCategory::Plan,
            plan_title(plan),
            plan_detail(plan),
        )),
        Body::ToolFailed(payload) => Some(Projection::informational(
            InboxCategory::ToolFailure,
            format!("Tool call {} failed", payload.call_id),
            Some(payload.error.clone()),
        )),
        Body::AgentCompleted(payload) => Some(Projection::informational(
            InboxCategory::Completion,
            payload
                .title
                .clone()
                .unwrap_or_else(|| "Agent completed".to_owned()),
            payload.reason.clone(),
        )),
        Body::AgentFailed(payload) => Some(Projection::informational(
            InboxCategory::ToolFailure,
            payload
                .title
                .clone()
                .unwrap_or_else(|| "Agent failed".to_owned()),
            payload.reason.clone(),
        )),
        Body::ContextUpdated(payload) if payload.percent_used >= CONTEXT_WARNING_PERCENT => {
            Some(Projection::informational(
                InboxCategory::ContextWarning,
                format!(
                    "Context window {}% used",
                    payload.percent_used.round() as i64
                ),
                Some(format!(
                    "{} of {} tokens used",
                    payload.used_tokens, payload.max_tokens
                )),
            ))
        }
        Body::UsageUpdated(payload)
            if payload
                .rate_limit_percent
                .is_some_and(|percent| percent >= RATE_LIMIT_WARNING_PERCENT) =>
        {
            let percent = payload.rate_limit_percent.unwrap_or_default();
            Some(Projection::informational(
                InboxCategory::RateLimit,
                format!("Rate limit {}% consumed", percent.round() as i64),
                payload.rate_limit_resets_at.map(|at| {
                    format!(
                        "resets at {}",
                        at.format(&time::format_description::well_known::Rfc3339)
                            .unwrap_or_else(|_| at.to_string())
                    )
                }),
            ))
        }
        _ => None,
    }
}

fn plan_title(plan: &Plan) -> String {
    let total = plan.steps.len();
    let completed = plan
        .steps
        .iter()
        .filter(|step| step.status == PlanStepStatus::Completed)
        .count();
    let base = plan.title.as_deref().unwrap_or("Plan");
    format!("{base} ({completed}/{total} steps complete)")
}

fn plan_detail(plan: &Plan) -> Option<String> {
    if plan.steps.is_empty() {
        return None;
    }
    Some(
        plan.steps
            .iter()
            .map(|step| {
                let marker = match step.status {
                    PlanStepStatus::Pending => "[ ]",
                    PlanStepStatus::InProgress => "[~]",
                    PlanStepStatus::Completed => "[x]",
                    PlanStepStatus::Cancelled => "[-]",
                };
                format!("{marker} {}", step.title)
            })
            .collect::<Vec<String>>()
            .join("\n"),
    )
}
